src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -2306,6 +2311,903 @@ internal unsafe void CM(B b)
+  IL_0006:  stloc.0
+  IL_0007:  newobj     ""C..ctor()""
+  IL_000c:  ldloc.0
+  IL_000d:  calli      0x4
+  IL_0012:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TypeArgumentNotSpecifiedNotInferred()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    static void M1<T>(int i) {}
+    static T M2<T>() => throw null;
+
+    static void M()
+    {
+        delegate*<int, void> ptr1 = &C.M1;
+        delegate*<string> ptr2 = &C.M2;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (9,38): error CS0411: The type arguments for method 'C.M1<T>(int)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         delegate*<int, void> ptr1 = &C.M1;
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "C.M1").WithArguments("C.M1<T>(int)").WithLocation(9, 38),
+                // (10,35): error CS0411: The type arguments for method 'C.M2<T>()' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         delegate*<string> ptr2 = &C.M2;
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "C.M2").WithArguments("C.M2<T>()").WithLocation(10, 35)
+            );
+        }
+
+        [Fact]
+        public void TypeArgumentSpecifiedOrInferred()
+        {
+            var verifier = CompileAndVerifyFunctionPointers(@"
+using System;
+unsafe class C
+{
+    static void M1<T>(T t) => Console.Write(t);
+    static void Main()
+    {
+        delegate*<int, void> ptr = &C.M1<int>;
+        ptr(1);
+        ptr = &C.M1;
+        ptr(2);
+    }
+}", expectedOutput: "12");
+
+            verifier.VerifyIL("C.Main()", expectedIL: @"
+{
+  // Code size       29 (0x1d)
+  .maxstack  2
+  .locals init (delegate*<int,void> V_0)
+  IL_0000:  ldftn      ""void C.M1<int>(int)""
+  IL_0006:  stloc.0
+  IL_0007:  ldc.i4.1
+  IL_0008:  ldloc.0
+  IL_0009:  calli      0x3
+  IL_000e:  ldftn      ""void C.M1<int>(int)""
+  IL_0014:  stloc.0
+  IL_0015:  ldc.i4.2
+  IL_0016:  ldloc.0
+  IL_0017:  calli      0x3
+  IL_001c:  ret
+}
+");
+        }
+
+        [Fact]
+        public void ReducedExtensionMethod()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe static class CHelper
+{
+    public static void M1(this C c) {}
+}
+unsafe class C
+{
+    static void M(C c)
+    {
+        delegate*<C, void> ptr1 = &c.M1;
+        delegate*<void> ptr2 = &c.M1;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (10,35): error CS8757: No overload for 'M1' matches function pointer 'delegate*<C,void>'
+                //         delegate*<C, void> ptr1 = &c.M1;
+                Diagnostic(ErrorCode.ERR_MethFuncPtrMismatch, "&c.M1").WithArguments("M1", "delegate*<C,void>").WithLocation(10, 35),
+                // (11,32): error CS8757: No overload for 'M1' matches function pointer 'delegate*<void>'
```

Comment:
>// (11,32): error CS8757: No overload for 'M1' matches function pointer 'delegate*<void>' [](start = 16, length = 89)

Could you please point me to the code that detects inapplicability for this scenario?

Reply:
BTW, It feels like the error is very confusing. It doesn't match either pointer, what else could it match?

---
In reply to: [397453464](https://github.com/dotnet/roslyn/pull/42488#discussion_r397453464) [](ancestors = 397453464)

Reply:
This occurs because when we attempt to convert the method group to a function pointer, we call `GetMethodGroupForFunctionPointerConversion`, which does not perform extension method lookup. I'll try to make this a bit clearer.

Reply:
> This occurs because when we attempt to convert the method group to a function pointer, we call ~~GetMethodGroupForFunctionPointerConversion~~ (Edit: GetMethodGroupFunctionPointerConversion), which does not perform extension method lookup.

Does it do this (does not perform extension method lookup) conditionally? Why the difference between two scenarios? We are referring to the same method, the target function pointer type is different. Something feels wrong/inconsistent here.

---
In reply to: [397562968](https://github.com/dotnet/roslyn/pull/42488#discussion_r397562968) [](ancestors = 397562968)